de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/project/FileSystemChangeListener.java

```diff
@@ -1,529 +1,870 @@
+                encoding);
+
+            editorManager.replaceAllEditorsForPath(oldFilePath, newFilePath);
+
+            annotationManager.updateAnnotationPath(oldFilePath.getFile(),
+                newFilePath.getFile());
+
+        } else if (newPathIsShared) {
+            //moved file into shared module
+            byte[] fileContent = getContent(oldFile);
+
+            SPath newFilePath = new SPath(newParentPath.getProject(),
+                newParentPath.getProjectRelativePath().append(relativePath));
+
+            activity = new FileActivity(user, Type.CREATED,
+                FileActivity.Purpose.ACTIVITY, newFilePath, null, fileContent,
+                encoding);
+
+            if (projectAPI.isOpen(oldFile)) {
+                Editor editor = projectAPI.openEditor(oldFile, false);
+
+                editorManager.addEditorMapping(newFilePath, editor);
+            }
+
+        } else if (oldPathIsShared) {
+            //moved file out of shared module
+            activity = new FileActivity(user, Type.REMOVED,
+                FileActivity.Purpose.ACTIVITY, oldFilePath, null, null, null);
+
+            editorManager.removeAllEditorsForPath(oldFilePath);
+
+            annotationManager.removeAnnotations(oldFilePath.getFile());
+
+        } else {
+            //neither source nor destination are shared
+            if (LOG.isTraceEnabled()) {
+                LOG.trace(
+                    "Ignoring non-shared resource move - resource: " + oldFile
+                        + ", old base parent: " + oldBaseParent
+                        + ", new base parent: " + newBaseParent);
+            }
 
-        if (project == null || !isValidProject(project) ||
-            !isCompletelyShared(project)) {
             return;
         }
 
-        path = makeAbsolutePathProjectRelative(path, project);
+        fireActivity(activity);
 
-        SPath spath = new SPath(project, path);
+        if (oldPathIsShared) {
+            if (fileIsOpen) {
+                EditorActivity closeOldEditorActivity = new EditorActivity(user,
+                    EditorActivity.Type.CLOSED, oldFilePath);
 
-        User user = resourceManager.getSession().getLocalUser();
-        IActivity activity;
+                fireActivity(closeOldEditorActivity);
+            }
 
-        byte[] bytes = new byte[0];
-        try {
-            bytes = virtualFileCopyEvent.getOriginalFile()
-                .contentsToByteArray();
-        } catch (IOException e) {
-            LOG.error("could not read content of original file "
-                + virtualFileCopyEvent.getOriginalFile(), e);
-            return;
+            //TODO reset the vector time for the old file
         }
 
-        activity = new FileActivity(user, Type.CREATED, Purpose.ACTIVITY, spath,
-            null, bytes, virtualFile.getCharset().name());
+        if (newPathIsShared) {
+            if (fileIsOpen) {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Merge this if statement with the enclosing one. ([squid:S1066](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1066))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=307&issues=AWcYB0JPkEeD8QRk19vQ&open=AWcYB0JPkEeD8QRk19vQ)

[ISSUE_KEY]: AWcYB0JPkEeD8QRk19vQ